### ISSUE TITLE
Declaration (.d.js) files

### ISSUE THREAD

**Issue by justinfagnani** [giltayar/proposal-types-as-comments#63]
I think it would be useful to specify a declaration-only subset of JS-with-type to use in declaration files. These would be `.d.js` files analogous to `.d.ts` files.

## Why
There are a few reasons to consider this:

### Performance

I've seen some comments about this proposal that it would replace the need to publish types with npm packages. This may be true some of the time (but see "Interoperability"), but there is a performance cost with having to parse the entire source rather than a declaration file.

Some large-scale TypeScript build systems, like in Google's internal monorepo, [use declaration files as a way to speed up compilation](http://neugierig.org/software/blog/2020/10/scaling-typescript.html).

When building a library, the system uses previously generated declaration files of dependencies as input to the compiler rather than the .ts source of those dependencies. This speeds up parsing and eliminates the need to perform inference (or even type-checking at all, if you trust the declaration files) for dependencies.

A declaration format would be usable by such build systems, and possibly able to be published with packages

### Interoperability

Declaration files remove inference from the type-checking equation, increasing interop. It's quite possible that some type-checker implementations will agree on the type-system _except_ inference, which is complicated to spec and tends to continuously evolve. Consider an alternate TypeScript type-checker written in Rust or Go for example, or simply different versions of TypeScript with different inference rules.

Relying on implementation source and inference, two type checkers may infer different return types for functions. Declaration files give a library a way to output the types with the inference algorithms the library used at the time of compilation, so that consuming type-checkers don't need to have compatible inference.

### External declarations

I assume that projects like DefinitelyTyped, and the ability to write your own type declarations for other libraries, will still exist for a indefinitely after proposal lands. First, because not all libraries will publish types. Second, because even if they do they might not use the type-system that a consumer requires. DefinitelyTyped would still exist for _TypeScript_ declarations, but you might have another type declaration project for a different type system. It'll need a declaration format.

Developers will still need a way to write declarations for dependencies as well.

## Why not .d.ts and libdef files?

Once we can write valid TypeScript or Flow types in standard JavaScript syntax, it would be an unfortunate context switch to have to read and write non-standard syntax in `.d.ts` and libdef files. It will be much better to converge on a standard declaration syntax, even for different type-systems, just as the main proposal converges on a standard types syntax.

Tools that wish to only support standard JS-with-types, and not TypeScript, Flow, etc., syntax, will also have declaration file needs, and this would give them a standard, interoperable format to use. New type-checkers could use a standard parser, for instance.

**Comment by ljharb** [comment]
Are you suggesting a third parse goal, alongside Script and Module?

If so, it would warrant a new file extension, like .mjs for Module (d.js is just a js extension; extensions don’t chain in a general sense), and similarly, the language wouldn’t be where that’s specified.

**Comment by justinfagnani** [comment]
Yes, and the extension could be something else like `.djs`. I realize the language wouldn't be where the extension is specified.

**Comment by DanielRosenwasser** [comment]
I think the only reason we would consider a `.d.js` file of some sort would be to enable non `type` imports that could be runtime validated; but I don't know how motivated I feel about that. I don't know if I feel convinced by the argument that there's a huge context switch between `.js` with comments and `.ts` (or `.d.ts`) that we would want to add a new format.

From the implementer side, I also am not a huge fan of the fact that `import "./foo.d.js"` becomes valid (it's not today, and you're supposed to just import from the `.js` file). It seems like that would add some complication.

**Comment by weswigham** [comment]
I don't get the point of this - an "ambient only subset file"? Why not just use the declaration format for whatever typesystem you're actually using? Since there's no associated semantics, there's no possibility of an "interoperable" format - they'll have have different syntactic subsets and meanings. Like, you can already see this today when you try to feed `.flow.js` files into TS itself; the subtle syntax and checking differences lead to failures more often than not.

Like this proposal doesn't introduce "standard syntax" - it reserves huge swaths of syntax space; there's still no actual standardization on behavior or interpretation within those.

**Comment by justinfagnani** [comment]
The proposal does introduce standard syntax in the form of a few new keywords and delimiters into the type space.

I guess one way of getting at the point is to ask about what TypeScript developers would be expected to do if this proposal succeeds. Would they write their code in .js files, but if they need to type external libraries write .d.ts files with subtly different type annotation syntax?

**Comment by weswigham** [comment]
Except it's not "subtly different" - it's expanded. TS would only support, in JS, the TS subset which is reserved by the JS spec (with existing jsdoc support covering the holes). Like, the spec's going to allow stuff like `type A = {I wrote an essay instead of a type}`; the JS syntax has _no meaning_ without the context of a greater typesystem. The syntax is a  venn diagram with overlap - what this proposal proposes JS reserves is a unique subset of syntax, separate from TS's syntax, and there's (significant) overlap, but not complete, meanwhile how that syntax is _interpreted_ by external typesystems, is wholly unspecified.

And runtimes have zero reason to import a file with only ambient declarations - they're for the typesystem itself.... trying to specify it is like trying to specify an unknown intermediate representation for an unknown decoder; there's no point. You don't need to specify anything - each typesystem will (or already does) adapt the available syntax space in a way that's convenient for their users.

**Comment by giltayar** [comment]
> I guess one way of getting at the point is to ask about what TypeScript developers would be expected to do if this proposal succeeds. Would they write their code in .js files, but if they need to type external libraries write .d.ts files with subtly different type annotation syntax?

Yes, write the code in `.js` files.  And, yes, for TS users that need to annotate existing no-typed code, they would still need the `.d.ts` files ans the existing mechanisms that support them.

As for "subtly different type annotation system": this is true today. The types in a `.d.ts` are subtly different than in a regular `.ts` file.
